lib/tasks/upgrade.rake

```diff
@@ -0,0 +1,56 @@
+# TRANSLATORS: do not translate
+desc <<-END_DESC
+  This task runs predefined upgrade steps.
+
+  Examples:
+    rake upgrade:run
+    rake upgrade:mark_as_ran TASK_NAME=some_rake_task,another_rake_task
+
+END_DESC
+
+namespace :upgrade do
+  task :run => :environment do
+    ENV['FOREMAN_UPGRADE'] = '1'
+    tasks = Foreman::Plugin.registered_plugins.values.map{|plugin| plugin.upgrade_tasks}.flatten
+    tasks.select!{|task| task.run?}
+
+    tasks.each_with_index do |task, index|
+      count = "#{index + 1}/#{tasks.count}"
+
+      message = if task.long_running?
+                  _("Upgrade Step %{count}: %{name}, This may take a long while.")
+                else
+                  _("Upgrade Step %{count}: %{name}")
+                end
+
+      puts '============================================='
+      puts message % {:name => task.name, :count => count}
+      task.mark_as_ran if run_task(task)
+    end
+  end
+
+  def run_task(task)
```

Comment:
I could see this logic living in the UpgradeTask itself

Reply:
Yeah, i kinda contemplated that, but running a rake task within a model seems incorrect.  To make it work from the model but not require it be run in a rake task, you would have to do some extra stuff like:

```
require 'rake'

Rake::Task.clear 
Foreman::Application.load_tasks
```

Rake tasks aren't really intended to be run from within the model